Unify Containers and MeshBlockPacks (#328)
parthenon-hpc-lab/parthenon: docs/interface/containers.md

> # MeshBlockDatas, MeshBlockData Iterators, and Variable Packs

Reviewer: ```suggestion
# MeshBlockData, MeshBlockData Iterators, and Variable Packs
```

> ## MeshBlockDatas

Reviewer: ```suggestion
## MeshBlockData
```

> A *meshblock data container* owns *variables.* Each variable is named, and the

Reviewer: Hm... lowercase meshblock, hard choice, but I think I have to go with:

```suggestion
A *mesh-block data container* owns *variables.* Each variable is named, and the
```

Reviewer 2: I don't know if I agree with this. I think `meshblock` has been the de facto standard both here and in the original Athena code.

Reviewer 3: I'm also in favor of keeping it to meshblock (or preferable even MeshBlock).

Reviewer: If you're going to spell it out as one concatenated word, then it should be `MeshBlock`, preferably with the code-styling ticks. I don't think "meshblock" is a common enough word to compound it without a hyphen.

Reviewer 3: I now went with using "`MeshBlockData`" everywhere so that it's consistent with the implementation.

Reviewer 2: My thinking was `meshblock` or `mb` are instances and `MeshBlock` is the class.

Reviewer 3: I don't have a strong opinion about the chosen style and am fine with the person with the strongest opinion just updating it.

> meshblock data container knows about various types of variables, such as whether the

Reviewer: The formal style guide outlined in my previous ~~tweet~~ review comment requires mesh-block now

```suggestion
mesh-block data container knows about various types of variables, such as whether the
```

variable is for cell-centered data, face-centered data, sparse data,
or dense data. (For more details on anonymous variables, see

> [here](Metadata.md).) Variables in a meshblock data container can be different

Reviewer: ```suggestion
[here](Metadata.md).) Variables in a mesh-block data container can be different
```

shapes, e.g., scalar, tensor, etc.A variable can be added to a

> meshblock data container as:

Reviewer: ```suggestion
mesh-block data container as:
```

```

### Automatic registration

For ease of use, the steps illustrated in the manual registration are automated in the
`mesh_data.GetOrAdd(string MeshBlockData_label, int partition_id)` function (e.g., used
in the `advection` example [here](../../example/advection/advection_driver.cpp) ).
Here, the partitioning in the background uses the default Mesh partition size (`pack_size`)
and the the total number of partition is accesse through `pmesh->DefaultNumPartitions();`.
Thus, a sample usage in a driver that executes Tasks on multiple partitions in parallel may
look like
```c++
TaskID no_dependency(0); // no dependency
const int num_partitions = pmesh->DefaultNumPartitions();
TaskRegion &single_tasklist_per_partition = tc.AddRegion(num_partitions);
for (int i = 0; i < num_partitions; i++) {
  auto &tl = single_tasklist_per_partition[i];
  // "base" MeshBlockData of blocks in partition i
  auto &mbase = pmesh->mesh_data.GetOrAdd("base", i);
  // MeshBlockData of the previous stage of blocks in partition i
  auto &mc0 = pmesh->mesh_data.GetOrAdd(stage_name[stage - 1], i);
  // MeshBlockData of the current stage of blocks in partition i

>   auto &mc1 = pmesh->mesh_data.GetOrAdd(stage_name[stage], i);

Reviewer 2: I like this.

shape = std::vector<int>({11,12,13,14});
```

> It is often desirable to extract from a meshblock data container a specific set of

Reviewer: ```suggestion
It is often desirable to extract from a mesh-block data container a specific set of
```

container.PackVariables(metadata, ids, map)

> container.PackVariables(metadata, map)

Reviewer 2: `data.*`?

Reviewer 3: Fixed.

container.PackVariables(metadata, ids, map)
container.PackVariables(metadata, map)
```

> If you would like all variables in a meshblock data container, 

Reviewer: ```suggestion
If you would like all variables in a mesh-block data container, 
```

For small kernels that perform little work (e.g., because of the simplicity of the kernel
itself or the small number of cells per `MeshBlock`, say 16<sup>3</sup> or smaller),
this can be a performance bottleneck when each kernel is launched per `MeshBlock`.
Parthenon therefore provides the capability to combine variables into a single data
structure that spans some number of meshblocks, the `MeshBlockPack`.

`MeshBlockPack`s created automatically and accessed transparently through `MeshData` objects.
These `MeshData` objects are stored as a `DataCollection` of shared pointers
in the `Mesh` object.

*IMPORTANT*, `MeshData` and `MeshBlockPack` are considered to be higher level
representations of lower level data, i.e., the data used in the simulation itself always
needs to be registered as `MeshBlockData` first before it can be accessed through
`MeshData` and `MeshBlockPacks`.

## Registering `MeshData`

`MeshData` is a lightweight object that aggregates multiple `MeshBlock`s.
Therefore, it needs to be setup/registered with some number of `MeshBlock`s
(at least one and at most all), which is referred to as partitioning.

> The `Partition` machinery found in [`utils/partition_stl_containers.hpp`](../../src/utils/partition_stl_containers.hpp).

Reviewer: Sentence fragment.

Reviewer 3: Fixed.

Registration and partitioning can be controlled manually or automatically
(recommended in multi-stage drivers).

### Manual registration

> The following steps (used in the `calculate_pi` example [here](../../example/calculate_pi/pi_driver.cpp) ) are need to manually register and fill a `MeshData` object.

Reviewer: ```suggestion
The following steps (used in the `calculate_pi` example [here](../../example/calculate_pi/pi_driver.cpp) ) are needed to manually register and fill a `MeshData` object.
```

```c++
// Number of MeshBlocks per Partition
const int pack_size = pmesh->DefaultPackSize();
// Partition all blocks of the Mesh into separate partitions containing pack_size MeshBlocks
auto partitions = partition::ToSizeN(pmesh->block_list, pack_size);
// Register a MeshData object for each partition (collection of blocks) using the partition
// number as label and containing references to the data stored in the "base" MeshBlockPack
for (int i = 0; i < partitions.size(); i++) {
  const std::string label = std::to_string(i);

>   auto mest_data = pmesh->mesh_data.Add(label);

Reviewer: ```suggestion
  auto mesh_data = pmesh->mesh_data.Add(label);
```

  // assign MeshBlocks of partitions[i] and data stored in "base" MeshBlockPack to MeshData object
  mesh_data->Set(partitions[i], "base");
}
```

There are two partitioning functions:
```C++
// Splits container into N equally sized partitions
template <typename T, typename Container_t>
Partition_t<T> ToNPartitions(Container_t<T> &container, const int N);

// Splits container into partitions of size N
template <typename T, typename Container_t>
std::vector<std::vector<T>> ToSizeN(Container_t<T> &container, const int N);
```
Both functions live within the namespace `parthenon::partition` and `Partition_t`
is defined as:
```C++
template<typename T>

> using Parition_t = std::vector<std::vector<T>>

Reviewer: ```suggestion
using Partition_t = std::vector<std::vector<T>>
```

Reviewer 2: Oops. This is my fault. It might actually be `Partition_t` in the code. Worth double-checking.

Reviewer 3: The code is correct. It was just the doc.
